gammapy/makers/background/reflected.py

```diff
@@ -179,74 +159,69 @@ def _region_angular_size(pixels, center):
-        self._pix_center = PixCoord.from_sky(self.center, geom.wcs)
-
-        # Make the ON reference map
-        mask = geom.region_mask([self.region]).data
-        # on_reference_map = WcsNDMap(geom=geom, data=mask)
-
-        # Extract all pixcoords in the geom
-        X, Y = geom.get_pix()
-        ONpixels = PixCoord(X[mask], Y[mask])
+    @lazyproperty
+    def exclusion_mask_ref(self):
+        """Exclusion mask reprojected"""
+        if self.exclusion_mask:
+            mask = self.exclusion_mask.interp_to_geom(self.geom_ref, fill_value=True)
+        else:
+            mask = WcsNDMap.from_geom(geom=self.geom_ref, data=1)
+            mask.data = mask.data.astype(bool)
+        return mask
 
+    @lazyproperty
+    def excluded_pix_coords(self):
+        """Excluded pix coords"""
         # find excluded PixCoords
-        mask = self.reference_map.data == 0
-        self.excluded_pixcoords = PixCoord(X[mask], Y[mask])
+        pix_y, pix_x = np.where(~self.exclusion_mask_ref.data)
+        return PixCoord(pix_x, pix_y)
 
+    @lazyproperty
+    def angle_min(self):
+        """Minimum angle"""
         # Minimum angle a region has to be moved to not overlap with previous one
-        min_ang = self._region_angular_size(ONpixels, self._pix_center)
-
         # Add required minimal distance between two off regions
-        self._min_ang = min_ang + self.min_distance
-
-        # Maximum possible angle before regions is reached again
-        self._max_angle = Angle("360deg") - self._min_ang - self.min_distance_input
+        return self.region_angular_size + self.min_distance
 
-    def find_regions(self):
-        """Find reflected regions."""
-        curr_angle = self._min_ang + self.min_distance_input
-        reflected_regions = []
+    @lazyproperty
+    def angle_max(self):
+        """Maximum angle"""
+        return Angle("360deg") - self.angle_min - self.min_distance_input
 
-        while curr_angle < self._max_angle:
-            test_reg = self._pix_region.rotate(self._pix_center, curr_angle)
-            if not np.any(test_reg.contains(self.excluded_pixcoords)):
-                region = test_reg.to_sky(self.reference_map.geom.wcs)
-                reflected_regions.append(region)
+    def reset_cache(self):
+        """Reset cached properties"""
+        for name, value in self.__class__.__dict__.items():
+            if isinstance(value, lazyproperty):
+                self.__dict__.pop(name, None)
 
-                curr_angle += self._min_ang
-                if self.max_region_number <= len(reflected_regions):
-                    break
-            else:
-                curr_angle = curr_angle + self.angle_increment
+    def run(self):
+        """Find reflected regions.
 
-        self.reflected_regions = reflected_regions
+        Returns
+        -------
+        regions : list of `SkyRegion`
+            Reflected regions
+        """
+        self.reset_cache()
```

Comment:
Does it need to reset the cache since the `RegionsFinder` is re-instantiated at each `run` of the `ReflectedRegionsBackgroundMaker`?
  You could also define getters and no setters for the `center` and `region` to force users to create a new `RegionFinder` for a new set of region and center position.

Reply:
Good point, indeed if `center` and `region` cannot be changed we can probably remove the reset cache pattern. I think in general the pattern is not bad, but I agree further restricting the attributes is even better...

Reply:
Looking more closely, it would require to declare all the attributes as non-settable attributes, which creates a lot of extra code. I the only way to further improve and simplify would be to move more code to `RegionGeom` as it already bundles the `region` and `wcs`. This way one could cache on the `RegionGeom` object.